tests/test_spark.py

```diff
@@ -0,0 +1,164 @@
+
+from gprofiler.log import get_logger_adapter
+from tests.conftest import build_image
+from tests.utils import wait_container_to_start
+
+# `BigDataSampler` receives a logger as an argument.
+logger = get_logger_adapter("gprofiler_spark_test")
+
+# List that includes all the metrics that are expected to be collected by `BigDataSampler` in SA mode.
+EXPECTED_SA_METRICS_KEYS = [
+    metric
+    for metrics_dict in (
+        SPARK_APPLICATION_GAUGE_METRICS,
+        SPARK_APPLICATION_DIFF_METRICS,
+        SPARK_AGGREGATED_STAGE_METRICS,
+        SPARK_EXECUTORS_METRICS,
+    )
+    for metric in metrics_dict.values()
+]
+
+SPARK_MASTER_HOST = "127.0.0.1"
+
+DISCOVER_INTERVAL_SECS = 5
+DISCOVER_TIMEOUT_SECS = 60
+
+
+@pytest.fixture
+def sparkpi_container(docker_client: DockerClient) -> Container:
+    """
+    This fixture is responsible for running SparkPi application in a container.
+    See `containers/spark/Dockerfile`
+    """
+    spark_image = build_image(docker_client=docker_client, runtime="spark")
+    container = docker_client.containers.run(
+        spark_image,
+        detach=True,
+        mounts=[],
+        network_mode="host",
+        pid_mode="host",
+        environment={"SPARK_MASTER_HOST": SPARK_MASTER_HOST},
+    )
+    wait_container_to_start(container)
+    try:
+        yield container
+    finally:
+        container.stop()
+        container.remove()
+
+
+def _wait_for_sparkpi_to_start() -> None:
+    """
+    Waits for SparkPi to be recognized by Master.
+    Doing so using `SparkRunningApps` class, `get_running_apps()` method.
+
+    Because Master is provisioning, it takes a few seconds until we receive a response from it.
+    """
+    running_apps = SparkRunningApps(SPARK_STANDALONE_MODE, f"http://{SPARK_MASTER_HOST}:8080", logger)
+    apps: Dict[str, Tuple[str, str]] = {}
+    while not apps:
+        try:
+            apps = running_apps.get_running_apps()
+        except requests.exceptions.ConnectionError as e:
+            if "Max retries exceeded" in str(e):
+                # Spark Master is not ready yet.
+                pass
+            sleep(1)
+    # We know that SparkPi is running, letting Master time to provision it.
+    sleep(5)
+
+
+def _validate_spark_sa_metricssnapshot(snapshot: MetricsSnapshot) -> None:
+    """
+    Validates that the snapshot contains all the expected metrics.
+    """
+    samples = snapshot.samples
+    assert len(samples) != 0, "No samples found in snapshot"
+    actual_metrics = {sample.name: sample.value for sample in samples}
+    # Creating a dict of the expected samples.
+    expected_metrics = dict.fromkeys(EXPECTED_SA_METRICS_KEYS, 0)
+    # Manually setting the expected values of some metrics.
+    expected_metrics["spark_aggregated_stage_active_stages"] = 1
+    expected_metrics["spark_num_applications_running"] = 1
+    expected_metrics["spark_job_diff_numTasks"] = 1000
+    expected_metrics["spark_job_numActiveStages"] = 1
```

Comment:
Instead of specifying them like that, add code like
```
if key not in ("spark_aggregated_stage_active_stages", ...):
    assert value == actual_metrics[key], ...
```
and explain in comment why you DON'T compare them.

Reply:
we spoke about that f2f

Reply:
Yeah it's my misunderstanding.